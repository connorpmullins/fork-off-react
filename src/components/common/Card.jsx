import React from "react";
import PropTypes from "prop-types";
import { colors, spacing, borderRadius, shadows } from "../../styles/theme";

const Card = ({
  children,
  padding = "4",
  elevation = "md",
  className = "",
  onClick,
  hoverable = false,
  ...props
}) => {
  const paddingMap = {
    2: "0.5rem",
    4: "1rem",
    6: "1.5rem",
    8: "2rem",
  };

  const elevationMap = {
    sm: "0 1px 2px 0 rgb(0 0 0 / 0.05)",
    md: "0 1px 3px 0 rgb(0 0 0 / 0.1), 0 1px 2px -1px rgb(0 0 0 / 0.1)",
    lg: "0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1)",
  };

  const baseStyles = {
    backgroundColor: colors.neutral.white,
    borderRadius: borderRadius.lg,
    padding: paddingMap[padding],
    boxShadow: elevationMap[elevation],
    transition: "all 150ms ease-in-out",
    cursor: onClick || hoverable ? "pointer" : "default",
    "&:hover": hoverable && {
      transform: "translateY(-2px)",
      boxShadow: shadows.lg,
    },
    ...props.style,
  };

  return (
    <div
      style={baseStyles}
      className={className}
      onClick={onClick}
      role={onClick ? "button" : "none"}
      tabIndex={onClick ? 0 : -1}
      {...props}
    >
      {children}
    </div>
  );
};

Card.propTypes = {
  children: PropTypes.node.isRequired,
  padding: PropTypes.oneOf(["2", "4", "6", "8"]),
  elevation: PropTypes.oneOf(["sm", "md", "lg"]),
  className: PropTypes.string,
  onClick: PropTypes.func,
  hoverable: PropTypes.bool,
};

export default Card;
