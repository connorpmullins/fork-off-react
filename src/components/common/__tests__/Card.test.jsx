import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Card from "../Card";

describe("Card", () => {
  it("matches snapshot with default props", () => {
    const { container } = render(
      <Card>
        <div>Test content</div>
      </Card>,
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with different padding sizes", () => {
    const { container, rerender } = render(
      <Card padding="2">
        <div>Content</div>
      </Card>,
    );
    expect(container).toMatchSnapshot();

    rerender(
      <Card padding="8">
        <div>Content</div>
      </Card>,
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with different elevations", () => {
    const { container, rerender } = render(
      <Card elevation="sm">
        <div>Content</div>
      </Card>,
    );
    expect(container).toMatchSnapshot();

    rerender(
      <Card elevation="lg">
        <div>Content</div>
      </Card>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders children correctly", () => {
    render(
      <Card>
        <div>Test content</div>
      </Card>,
    );
    expect(screen.getByText("Test content")).toBeInTheDocument();
  });

  it("applies different padding sizes", () => {
    const { rerender } = render(
      <Card padding="2" data-testid="card">
        <div>Content</div>
      </Card>,
    );
    expect(screen.getByTestId("card")).toHaveStyle({ padding: "0.5rem" });

    rerender(
      <Card padding="8" data-testid="card">
        <div>Content</div>
      </Card>,
    );
    expect(screen.getByTestId("card")).toHaveStyle({ padding: "2rem" });
  });

  it("applies different elevation styles", () => {
    const { rerender } = render(
      <Card elevation="sm" data-testid="card">
        <div>Content</div>
      </Card>,
    );
    expect(screen.getByTestId("card")).toHaveStyle({
      boxShadow: "0 1px 2px 0 rgb(0 0 0 / 0.05)",
    });

    rerender(
      <Card elevation="lg" data-testid="card">
        <div>Content</div>
      </Card>,
    );
    expect(screen.getByTestId("card")).toHaveStyle({
      boxShadow:
        "0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1)",
    });
  });

  it("handles click events when provided", () => {
    const handleClick = jest.fn();
    render(
      <Card onClick={handleClick}>
        <div>Clickable content</div>
      </Card>,
    );

    fireEvent.click(screen.getByRole("button"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("applies hover styles when hoverable prop is true", () => {
    render(
      <Card hoverable data-testid="card">
        <div>Hoverable content</div>
      </Card>,
    );

    expect(screen.getByTestId("card")).toHaveStyle({ cursor: "pointer" });
  });

  it("applies correct accessibility attributes when clickable", () => {
    render(
      <Card onClick={() => {}}>
        <div>Clickable content</div>
      </Card>,
    );

    expect(screen.getByRole("button")).toHaveAttribute("tabIndex", "0");
  });
});
