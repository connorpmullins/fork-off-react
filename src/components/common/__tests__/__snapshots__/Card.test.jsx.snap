// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches snapshot with default props 1`] = `
<div>
  <div
    class=""
    role="none"
    style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 1rem; box-shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1), 0 1px 2px -1px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
    tabindex="-1"
  >
    <div>
      Test content
    </div>
  </div>
</div>
`;

exports[`Card matches snapshot with different elevations 1`] = `
<div>
  <div
    class=""
    role="none"
    style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 1rem; box-shadow: 0 1px 2px 0 rgb(0 0 0 / 0.05); transition: all 150ms ease-in-out; cursor: default;"
    tabindex="-1"
  >
    <div>
      Content
    </div>
  </div>
</div>
`;

exports[`Card matches snapshot with different elevations 2`] = `
<div>
  <div
    class=""
    role="none"
    style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 1rem; box-shadow: 0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
    tabindex="-1"
  >
    <div>
      Content
    </div>
  </div>
</div>
`;

exports[`Card matches snapshot with different padding sizes 1`] = `
<div>
  <div
    class=""
    role="none"
    style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 0.5rem; box-shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1), 0 1px 2px -1px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
    tabindex="-1"
  >
    <div>
      Content
    </div>
  </div>
</div>
`;

exports[`Card matches snapshot with different padding sizes 2`] = `
<div>
  <div
    class=""
    role="none"
    style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 2rem; box-shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1), 0 1px 2px -1px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
    tabindex="-1"
  >
    <div>
      Content
    </div>
  </div>
</div>
`;
